### PR TITLE
Cr 979

### PR DESCRIPTION
**Pull request by andrewjohnysons**
Motivation and Context
Get case by UPRN endpoint should only return the latest case for that UPRN with a valid address

What has changed
modified endpoint to return a single case or return a 404
Service now returns a single case to the endpoint or throws a CTPException

How to test?
Unit tests have been amended to cope with the changes and there are new Cucumber tests which test for the latest valid case to be returned

Links

### CODE REVIEW

**Comment by andrewjohnysons** [comment]
Passed in DEV

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java [hunk 1]
```diff
@@ -76,7 +81,7 @@ public CaseDTO modifyAddress(final AddressChangeDTO addressChanges) throws CTPEx
 
     Optional<CollectionCase> caseMatch = dataRepo.readCollectionCase(caseId);
 
-    if (!caseMatch.isPresent()) {
+    if (caseMatch.isEmpty()) {
```
Comment (PeterBochel) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:84]: That reads much better with isEmpty :)

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java [hunk 2]
```diff
@@ -48,25 +49,29 @@
...
-
-    return caseData;
+            .filter(c -> !c.getCaseType().equals(CaseType.HI.name()))
+            .filter(c -> !c.isAddressInvalid())
+            .max(Comparator.comparing(CollectionCase::getCreatedDateTime));
+
+    if (result.isPresent()) {
+      log.with("case", result.get().getId())
+          .with("uprn", uprnValue)
+          .debug("HH latest case retrieved for UPRN");
```
Comment (PeterBochel) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:69]: Minor, but this is now incorrect? Can it not be a CE or SPG?

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java [hunk 3]
```diff
@@ -48,25 +49,29 @@
...
-            .filter(c -> c.getCaseType().equals(CaseType.HH.name()))
-            .collect(Collectors.toList());
-    List<CaseDTO> caseData = mapperFacade.mapAsList(results, CaseDTO.class);
-
-    log.with("cases", caseData.size())
-        .with("uprn", uprnValue)
-        .debug("HH case(s) retrieved for UPRN");
-
-    return caseData;
+            .filter(c -> !c.getCaseType().equals(CaseType.HI.name()))
```
Comment (PeterBochel) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:62]: Why the the change from equals HH to not-equals HI ? 
I'm sure you have a good reason but I just can't see it

Comment (andrewjohnysons) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:62]: Added comment to README file

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java [hunk 4]
```diff
@@ -31,27 +30,20 @@
   @Autowired private CaseService caseService;
 
   /**
-   * the GET end point to get a List of HH Cases by UPRN
+   * the GET end point to return the latest valid HH Case by UPRN
    *
    * @param uprn the UPRN
-   * @return List of returned HH cases for the UPRN
```
Comment (andrewjohnysons) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/endpoint/CaseEndpoint.java:37]: Fixed these HH messages

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/CaseService.java [hunk 5]
```diff
@@ -17,10 +16,11 @@
    * Retrieve the data relating to HH Cases by address UPRN
    *
    * @param uprn of address for which HH case details are requested
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/CaseService.java:18]: javadoc lines need updating.

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java [hunk 6]
```diff
@@ -86,10 +84,9 @@ public void getHHCaseByUPRNFound() throws Exception {
 
     CollectionCase hhCase = this.collectionCase.get(0);
 
-    List<CaseDTO> caseDTO = caseSvc.getHHCaseByUPRN(UPRN);
-    CaseDTO rmCase = caseDTO.get(0);
+    CaseDTO rmCase = caseSvc.getLatestValidNonHICaseByUPRN(UPRN);
 
-    assertThat(caseDTO, hasSize(1));
+    assertNotNull(rmCase);
```
Comment (rpcatling) [inline src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java:89]: rename various test method names to match your new service method name

Comment (rpcatling) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java:89]: It is also not clear to me that we have tested the max by createdDateTime functionality.  We need sample data with different created date times, and maybe an extra test method or 2  to ensure this is working as implemented.

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java [hunk 7]
```diff
@@ -48,25 +49,29 @@
...
+            .filter(c -> !c.isAddressInvalid())
```
Comment (rpcatling) [inline src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:63]: this line needs more code coverage: need to test with valid and invalid address

File: README.md [hunk 8]
```diff
@@ -411,7 +411,10 @@ This example uses a case uuid of: f868fcfc-7280-40ea-ab01-b173ac245da3
     # Nicely close down rabbit connection (Probably not required)
     http --auth generator:hitmeup GET "http://localhost:8171/rabbit/close"
     
-
+Modified enpoint /cases/uprn/ - method name changed from
```
Comment (rpcatling) [inline README.md:414]: typo for "endpoint"

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java (hunk 6 above)
Comment (philwhiles) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java:89]: I just spotted this also Rob - no unit test coverage with more than case to assert the correct filtering. As there is duplicated functionality in the CaseServceImpl and an existing JIRA ticket to remove this duping, I have agreed with Andy that he will pick that other ticket up immediately after this one. In doing the de-dupe refactoring he can move tests that assert the same functionality in UACServiceImpl to test the de-duped method. Adding new unit tests now that will be removed is unnecessary

File: src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java (hunk 2 above)
Comment (andrewjohnysons) [reply src/main/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImpl.java:69]: Fixed the debug message

**Comment by andrewjohnysons** [comment]
passes in dev

File: src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java (hunk 6 above)
Comment (andrewjohnysons) [reply src/test/java/uk/gov/ons/ctp/integration/rhsvc/service/impl/CaseServiceImplTest.java:91]: Addressed HH issues and added unit tests for CaseServiceImpl -> tests latest and valid address and non HI cases

**Comment by andrewjohnysons** [comment]
Successful run in dev

**Pull request merged**